### PR TITLE
Enable state probing for OptimizationSolver on CPU and Loihi backend

### PR DESCRIPTION
**Pull request by AlessandroPierro**
Issue Number:

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-optimization/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [x] Feature

## What is the new behavior?

- It is now possible to track the neuron state through time during the execution of OptimizationSolver and obtain the timeseries at the end of execution
- The state probing can be requested setting `SolverConfig.probe_state` to `True` when calling the `solve` method of `OptimizationSolver` 
-  The state timeseries is returned in `SolverReport.state_timeseries`

## Does this introduce a breaking change?

- [x] No

### CODE REVIEW

**Comment by AlessandroPierro** [comment]
Ready for review. Tests are passing locally, and will pass in the CI when https://github.com/lava-nc/lava/pull/675 is merged.

**Review by GaboFGuerra — APPROVED** [review]
Thanks Alessandro!

File: src/lava/lib/optimization/solvers/generic/monitoring_processes/solution_readout/models.py [hunk 1]
```diff
@@ -42,26 +42,44 @@ def run_spk(self):
...
             self.solution_step = abs(timestep)
+            self.solution[:] = self._decode_solution(raw_solution)
             self.min_cost[:] = np.asarray([cost[0], min_cost_id])
             if cost[0] < 0:
-                print(
-                    f"Host: better solution found by network {min_cost_id} at "
-                    f"step {abs(timestep)-2} "
-                    f"with cost {cost[0]}: {self.solution}"
-                )
+                self._printout_new_solution(cost, min_cost_id, timestep)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/monitoring_processes/solution_readout/models.py:51]: Nice cleanup!

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 2]
```diff
@@ -308,15 +328,37 @@ def _get_requirements_and_protocol(
...
+        """
+        cost_timeseries = self._get_probed_data(
+            tracker=self._cost_tracker, var_name="optimality"
+        )
+        state_timeseries = self._get_probed_data(
+            tracker=self._state_tracker, var_name="variable_assignment"
+        )
+        if state_timeseries is not None:
+            state_timeseries &= 0x1F
+            state_timeseries = state_timeseries.astype(np.int8) >> 4
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solver.py:350]: Is this dublicate code? This transformation seems to happen in solution_readout/models.py as well. Is it necessary here again? Seems too low level for me for solver.py.

Comment (phstratmann) [reply src/lava/lib/optimization/solvers/generic/solver.py:350]: Hm...
Adding such low-level detail here looks very error prone to me. Let's assume we change something about the solver (e.g. not the 5th, but the 8th position is suddenly the value at cost readout). We will probably forget to change it here.
Thus, maybe we can introduce a util function that translates between read states and the value; or an enum that defines at what position the value is. What do you think?

Comment (AlessandroPierro) [reply src/lava/lib/optimization/solvers/generic/solver.py:350]: Where would you place this function? What about in the models.py file of SolutionReadout? We already a member function there with the same purpose, should we make it a static method and re-use it in solver.py?

**Pull request merged**